cpp/src/arrow/scalar.h

```diff
@@ -111,6 +111,12 @@ struct ARROW_EXPORT Scalar : public std::enable_shared_from_this<Scalar>,
  protected:
   Scalar(std::shared_ptr<DataType> type, bool is_valid)
       : type(std::move(type)), is_valid(is_valid) {}
+
+  // 16 bytes of scratch space to enable ArraySpan to be a view onto any
+  // Scalar- including binary scalars where we need to create a buffer
+  // that looks like two 32-bit or 64-bit offsets.
+  alignas(int64_t) std::array<uint8_t, sizeof(int64_t) * 2> scratch_space_;
```

Comment:
Do we want to put this in the base `Scalar` or only in an intermediate subclass `ScalarWithScratchSpace` perhaps?

Reply:
If you'd prefer, the intermediate class wouldn't be hard. I'd think leaving the scratch space here is fine until we have a reason to move it